### PR TITLE
Fix Field Indexer and get_element_indices for layered solids and add example of a thermal analysis

### PR DESCRIPTION
**Pull request by roosre**
While adding an example for a thermal analysis (examples/013_thermal_example.py), several issues were found in the result extraction for layered solid elements where the field indexers are used. The main issue is that fields which have one entry per entity behave slightly different if compared with one that has multiple entries per entity. This became visible because a solid model with "analysis ply-wise" extrusion was used. So there is only one entry per entity for the layered fields such as angles, materials etc.

Here are some details:
- `FieldIndexer` has to handle cases where the data pointers of a field are set in DPF or not. This is a performance optimization in DPF. The data pointers are only available if multiple entries are set per entity. This is addressed by implementing helper functions `get_property_field_indexer` and `get_field_indexer` which automatically returns the correct indexer.
- The Protocol of the (Property)FieldIndexer was enriched by `by_id_as_array` to clearly distinguish between a single value and an array of values because the return type was dependent on if data pointers are available or not.
- An issue with the DPF container/server 2024 R1 pre0 on extracting laminate offsets is fixed. This did not show up before because the `FieldIndexer` which returns a single value was used explicitly. `get_field_indexer` returns now the correct indexer but the bug in the backend becomes visible now.
- By accident, the `number of corner nodes` was used instead of `number of nodes per spot` when computing the indices. This makes a difference if layered solid elements are processed. See `get_selected_indices` in src/ansys/dpf/composites/select_indices.py.

### CODE REVIEW

File: src/ansys/dpf/composites/_indexer.py [hunk 1]
```diff
@@ -66,6 +66,35 @@ def by_id(self, entity_id: int) -> Optional[NDArray[np.int64]]:
         """Get index by id."""
 
 
+def _has_data_pointer(field: PropertyField | Field) -> bool:
```
Comment (greschd) [inline src/ansys/dpf/composites/_indexer.py:69]: The `X | Y` syntax for types is a Python 3.10 feature. It's probably easiest to just use `Union[X, Y]` for now. Alternatively, we can first drop 3.9 support.

**Comment by greschd** [comment]
General question: are we planning to backport these fixes to the `0.6` release branch?

File: src/ansys/dpf/composites/_indexer.py [hunk 2]
```diff
@@ -80,56 +114,31 @@ class PropertyFieldIndexerNoDataPointer:
     """Indexer for a property field with no data pointer."""
 
     def __init__(self, field: PropertyField):
```
Comment (roosre) [inline src/ansys/dpf/composites/_indexer.py:116]: The implements are ordered now. All PropertyFieldIndexerfirst, then the FieldIndexer classes.

File: src/ansys/dpf/composites/_indexer.py [hunk 3]
```diff
@@ -252,39 +256,90 @@ def by_id(self, entity_id: int) -> Optional[NDArray[np.double]]:
         )
 
 
-class PropertyFieldIndexerWithDataPointerNoBoundsCheck:
-    """Indexer for a property field with data pointer and no bounds checks."""
+# DPF does not set the data pointers if a field has just
+# one value per entity. Therefore, it is unknown if
+# data pointer are set for some field of layered elements, for
+# instance angles, thickness etc.
+def get_field_indexer(field: Field) -> FieldIndexSingleValueProtocol:
```
Comment (roosre) [inline src/ansys/dpf/composites/_indexer.py:263]: New function to get the correct FieldInder. This function and `get_property_field_indexer` should be used from externally only.

File: src/ansys/dpf/composites/_indexer.py [hunk 4]
```diff
@@ -66,6 +66,40 @@ def by_id(self, entity_id: int) -> Optional[NDArray[np.int64]]:
...
 
+def _has_data_pointer(field: PropertyField | Field) -> bool:
+    if (
+        field._data_pointer is not None and field._data_pointer.any()
+    ):  # pylint: disable=protected-access
+        return True
+    return False
+
+
+def get_property_field_indexer(
```
Comment (roosre) [inline src/ansys/dpf/composites/_indexer.py:77]: This function was implemented in _layup_info.py before.

File: src/ansys/dpf/composites/_indexer.py [hunk 5]
```diff
@@ -252,39 +256,90 @@ def by_id(self, entity_id: int) -> Optional[NDArray[np.double]]:
...
+        return FieldIndexerWithDataPointer(field)
+    return FieldIndexerNoDataPointer(field)
+
+
+class FieldIndexerNoDataPointer:
+    """Indexer for a dpf field with no data pointer."""
+
+    def __init__(self, field: Field):
         """Create indexer and get data."""
         if field.scoping.size > 0:
```
Comment (roosre) [inline src/ansys/dpf/composites/_indexer.py:280]: Check if size is 0 was missing.

File: tests/element_info_output_all_element_types_test.py [hunk 6]
```diff
@@ -288,3 +297,64 @@ def get_layup_info_for_rst(rst_file):
         assert str(exc_info.value).startswith(
             "Spot index 2 is greater or equal to the number of spots"
         )
+
+
+def test_select_indices_all_element_types(dpf_server):
```
Comment (roosre) [inline tests/element_info_output_all_element_types_test.py:302]: New test to verify the bug fixes.

**Comment by roosre** [comment]
> General question: are we planning to backport these fixes to the `0.6` release branch and 2024R2 `pyansys` metapackage?

Hi @greschd , yes. That would make sense from my point of view.

File: src/ansys/dpf/composites/_indexer.py [hunk 7]
```diff
@@ -52,18 +52,52 @@ def setup_index_by_id(scoping: Scoping) -> IndexToId:
     return IndexToId(mapping=indices, max_id=len(indices) - 1)
 
 
-class PropertyFieldIndexerSingleValue(Protocol):
+class PropertyFieldIndexerProtocol(Protocol):
     """Protocol for single value property field indexer."""
 
     def by_id(self, entity_id: int) -> Optional[np.int64]:
```
Comment (janvonrickenbach) [inline src/ansys/dpf/composites/_indexer.py:58]: Maybe mention that this function can throw if there is multiple values per entitiy

**Pull request merged**